modules/wolfgame.py

```diff
@@ -787,6 +798,249 @@ def fpinger(cli, nick, chan, rest):
+            if value not in var.PING_IF_NUMS.keys():
+                var.PING_IF_NUMS[value] = []
+            var.PING_IF_NUMS[value].append(cloak)
+            if old is not None:
+                if old in var.PING_IF_NUMS.keys():
+                    if cloak in var.PING_IF_NUMS[old]:
+                        var.PING_IF_NUMS[old].remove(cloak)
+
+def join_timer_handler(cli):
+    while var.PHASE == "join":
+        with var.GRAVEYARD_LOCK:
+            if var.PING_JOIN_TIMER and var.PING_JOIN_TIMER + timedelta(seconds=10) <= datetime.now():
+                to_ping = []
+                pl = var.list_players()
+
+                var.PINGING_IFS = True
+                checker = []
+                chk_acc = []
+
+                for num in var.PING_IF_NUMS_ACCS:
+                    if num <= len(pl):
+                        chk_acc.extend(var.PING_IF_NUMS_ACCS[num])
+
+                if not var.ACCOUNTS_ONLY:
+                    for num in var.PING_IF_NUMS:
+                        if num <= len(pl):
+                            checker.extend(var.PING_IF_NUMS[num])
+
+                for acc in chk_acc[:]:
+                    if acc in var.PINGED_ALREADY_ACCS:
+                        chk_acc.remove(acc)
+
+                for cloak in checker[:]:
+                    if cloak in var.PINGED_ALREADY:
+                        checker.remove(cloak)
+
+                if not chk_acc and not checker:
+                    var.PINGING_IFS = False
+                    var.PING_JOIN_TIMER = 0
+                    return
+
+                @hook("whospcrpl", hookid=387)
+                def ping_altpingers(cli, server, nick, ident, cloak, _, user, status, acc):
+                    if ('G' in status or '+' in status or is_user_stasised(user)[0] or
+                            not var.PINGING_IFS or user == botconfig.NICK or user in pl):
+
+                        return
+
+                    if acc and acc != "*":
+                        if acc in chk_acc and acc in var.PING_PREFS_ACCS and var.PING_PREFS_ACCS[acc] in ("once", "all"):
+                            to_ping.append(user)
+                            var.PINGED_ALREADY_ACCS.append(acc)
+
+                    elif not var.ACCOUNTS_ONLY and cloak in checker and cloak in var.PING_PREFS and var.PING_PREFS[cloak] in ("once", "all"):
+                        to_ping.append(user)
+                        var.PINGED_ALREADY.append(cloak)
+
+                @hook("endofwho", hookid=387)
+                def fetch_altpingers(*stuff):
+                    var.PINGING_IFS = False
+                    var.PING_JOIN_TIMER = 0
+                    decorators.unhook(HOOKS, 387)
+                    if to_ping:
+                        to_ping.sort(key=lambda x: x.lower())
+                        cli.msg(botconfig.CHANNEL, "PING! {0} players! {1}".format(len(pl), " ".join(to_ping)))
+
+                cli.who(botconfig.CHANNEL, "%nushaf")
+
+        time.sleep(5)
```

Comment:
This is still horribly done. A better construction would be instead of having a dedicated thread for this, to set a 10-second timer on each !join that calls the handler function. Each subsequent join would then cancel the current timer and set a new one for 10 seconds, to ensure that the ping always happens 10 seconds after the latest join.